### PR TITLE
bpo-46480: add typing.assert_type

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Another pending typing-sig proposal. This function serves as a static assertion to the type checker about the type of a value.

https://bugs.python.org/issue46480

### CODE REVIEW

File: Doc/library/typing.rst
Comment (sobolevn) [inline Doc/library/typing.rst:2117]: My 2c is that this should go somewhere at the bottom of the page. Why? Because it is only interesting for advanced users (type-checker authors / stubs maintainers / libs authors).

It might confuse people a lot if it is "advertised" in the same section as regular typing features.

What do others think?

Comment (sobolevn) [reply Doc/library/typing.rst:2117]: Maybe something like "Types maintenance helpers"? Or similar?

Comment (JelleZijlstra) [reply Doc/library/typing.rst:2117]: We already have the similarly obscure `@type_check_only` in this section, along with `@no_type_check_decorator` which I'm pretty sure nobody uses. I could get on board with moving it near the bottom of the section, though.

**Comment by JelleZijlstra** [comment]
@gvanrossum this one is also ready I believe. Feedback on this function in https://mail.python.org/archives/list/typing-sig@python.org/thread/MITFQ6Z45RRMXY3HNM66IC3XXS3TA3JN/#MITFQ6Z45RRMXY3HNM66IC3XXS3TA3JN was positive. (I also proposed assert_error() there, but dropped it because of negative feedback.)

File: Doc/library/typing.rst
Comment (davidfstr) [inline Doc/library/typing.rst]: Maybe clarify that there is no other _runtime_ effect (such as checking the value's type):

* At runtime this returns the first argument unchanged **and otherwise does nothing.**

File: Lib/test/test_typing.py
Comment (davidfstr) [inline Lib/test/test_typing.py]: Probably want to include `self.assertIs(...)` checks around these calls too? (Presumably each call returns `42`.)

File: Lib/typing.py
Comment (davidfstr) [inline Lib/typing.py]: (Ditto, RE my related comment in the RST documentation file)

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Good point, done

File: Lib/test/test_typing.py
Comment (JelleZijlstra) [reply Lib/test/test_typing.py]: Done

File: Lib/typing.py
Comment (JelleZijlstra) [reply Lib/typing.py]: Done, thanks for the review!

**Comment by davidfstr** [comment]
It occurs to me that users of "assert_type()" may be surprised that it _doesn't_ actually perform an assertion at runtime.

Consider code like:

```
from typing import assert_type

def parse_int(value) -> int:
    # Hmm. A casual observer would likely read this incorrectly as `assert isinstance(value, str)`...
    assert_type(value, str)
    return int(value)
```

I wonder if we might consider a word other than "assert", or maybe add the word "static" somewhere. Suggestions:

* `prove_type(value, str)`
* `static_assert_type(value, str)`

**Comment by JelleZijlstra** [comment]
That's a reasonable concern, but it's probably better discussed on the typing-sig thread. I'm happy to change the PR if there's consensus for another name. I'll note though that `cast()` has the same potential problem, and I haven't heard of confusion because of that.

**Comment by davidfstr** [comment]
> it's probably better discussed on the typing-sig thread

I'll plan to make a post tomorrow.

**Comment by JelleZijlstra** [comment]
@gvanrossum I think this should be ready to merge soon, since people on the typing-sig thread now seem on board with the `assert_type` name. I'd appreciate some more reviews for the documentation though, maybe @Fidget-Spinner or @AlexWaygood.

**Comment by AlexWaygood** [comment]
> @gvanrossum I think this should be ready to merge soon, since people on the typing-sig thread now seem on board with the `assert_type` name. I'd appreciate some more reviews for the documentation though, maybe @Fidget-Spinner or @AlexWaygood.

I'll try to take a look soon!

File: Doc/library/typing.rst
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
   At runtime this returns the first argument unchanged with no side effects.
```

File: Doc/library/typing.rst
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
   Assert (to the type checker) that *value* has a revealed type of *typ*.
```

The type checker has no knowledge of the _actual type_ of `value`. It only has knowledge of its _inferred type_ of `value`, which is what this function tests.

File: Doc/library/typing.rst
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
           assert_type(name, str)  # OK, revealed type of `name` is `str`
```

File: Doc/library/typing.rst
Comment (AlexWaygood) [inline Doc/library/typing.rst:2139]:
```suggestion

   This function is useful for ensuring the type checker's understanding of a
   script is in line with the developer's intentions::

       def complex_function(arg: object):
           # Do some complex type-narrowing logic,
           # after which we hope the inferred type will be `int`
           ...
           # Test whether the type checker correctly understands our function
           assert_type(arg, int)

```

(Or some other example along these lines)

**Review by AlexWaygood — COMMENTED** [review]
Thanks! Looks pretty good, just a few thoughts :)

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: I don't think "revealed type" makes sense if we're not calling reveal_type():

```suggestion
   Assert (to the type checker) that *value* has an inferred type of *type*.
```

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]:
```suggestion
           assert_type(name, str)  # OK, inferred type of `name` is `str`
```

**Comment by JelleZijlstra** [comment]
Thanks for the review!

**Review by AlexWaygood — APPROVED** [review]
LGTM!

File: Lib/typing.py
Comment (davidfstr) [inline Lib/typing.py]: Nit: Unusual to have a blank line preceding a closing """. Suggest removing it.

**Review by davidfstr — COMMENTED** [review]
Just 1 tiny nit:

File: Lib/typing.py
Comment (JelleZijlstra) [reply Lib/typing.py]:
```suggestion
    """
```

Comment (JelleZijlstra) [reply Lib/typing.py]: Thanks!

**Comment by JelleZijlstra** [comment]
Planning to merge this tomorrow unless I get more feedback.

File: Doc/library/typing.rst
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
   Assert (to the type checker) that *value* has an inferred type of *typ*.
```
The parameter name is `typ`, not `type` :)

Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Good catch! Also, the other param is named `val`, not `value`. I chose those names because `cast()` uses them too.
```suggestion
   Assert (to the type checker) that *val* has an inferred type of *typ*.
```

**Pull request merged**
